core/drivers/crypto/hisilicon/hpre_montgomery.c

```diff
@@ -0,0 +1,510 @@
+	key->priv = calloc(1, key_size);
+	if (!key->priv)
+		goto priv_err;
+
+	key->pub = calloc(1, key_size);
+	if (!key->pub)
+		goto pub_err;
+
+	return TEE_SUCCESS;
+pub_err:
+	free(key->priv);
+	key->priv = NULL;
+priv_err:
+	EMSG("HPRE montgomery alloc key pair fail");
+
+	return TEE_ERROR_OUT_OF_MEMORY;
+}
+
+static TEE_Result gen_random_privkey(uint8_t *priv, size_t key_bits)
+{
+	size_t key_size = BITS_TO_BYTES(key_bits);
+	TEE_Result ret = TEE_SUCCESS;
+
+	if (!priv) {
+		EMSG("Privkey param is NULL");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	ret = hw_get_random_bytes(priv, key_size);
+	if (ret) {
+		EMSG("Fail to fill privkey");
+		return TEE_ERROR_NO_DATA;
+	}
+
+	return ret;
+}
+
+static enum hisi_drv_status
+hpre_montgomery_params_alloc(struct hpre_montgomery_msg *msg)
+{
+	uint32_t size = HPRE_MONTGOMERY_TOTAL_BUF_SIZE(msg->key_bytes);
+
+	msg->key = calloc(1, size);
+	if (!msg->key) {
+		EMSG("Fail to alloc montgomery key buf");
+		return HISI_QM_DRVCRYPT_ENOMEM;
+	}
+
+	msg->key_dma = virt_to_phys(msg->key);
+	msg->in = msg->key + HPRE_X_KEY_SIZE(msg->key_bytes);
+	msg->in_dma = msg->key_dma + HPRE_X_KEY_SIZE(msg->key_bytes);
+	msg->out = msg->in + msg->key_bytes;
+	msg->out_dma = msg->in_dma + msg->key_bytes;
+
+	return HISI_QM_DRVCRYPT_NO_ERR;
+}
+
+static void hpre_montgomery_params_free(struct hpre_montgomery_msg *msg)
+{
+	if (msg->key) {
+		memzero_explicit(msg->key, HPRE_X_KEY_SIZE(msg->key_bytes));
+		free(msg->key);
+		msg->key = NULL;
+	}
+}
+
+static enum hisi_drv_status
+hpre_montgomery_params_pretreatment(struct hpre_montgomery_msg *msg)
+{
+	uint8_t *p = msg->key;
+	uint8_t *a = p + msg->key_bytes;
+	uint8_t *k = a + msg->key_bytes;
+	uint8_t *u = msg->in;
+	uint8_t *dst = k;
+	uint32_t bsize = msg->key_bytes;
+	uint32_t dsize = msg->curve_bytes;
+	uint32_t psize = msg->curve_bytes;
+	uint32_t shift = 0;
+	int ret = 0;
+
+	/*
+	 * This is a pretreatment of X25519 with a 32-byte integer,
+	 * as described in RFC 7748:
+	 * Set the three LSB of the first byte and MSB of the last
+	 * to zero, set the second MSB of the last byte to 1.
+	 * When receiving u-array, set MSB of last byte to zero.
+	 * HPRE hardware module uses big-endian mode, so the bytes to be
+	 * set are reversed compared to RFC 7748
+	 */
+	if (msg->key_bytes == BITS_TO_BYTES(X25519_KEY_BITS)) {
+		dst[31] &= 0xF8;
+		dst[0] &= 0x7F;
+		dst[0] |= 0x40;
+		u[0] &= 0x7F;
+	} else {
+		/*
+		 * This is a pretreatment of X448 with a 56-byte integer,
+		 * as described in RFC 7748:
+		 * For X448, set the two LSB of the first byte to 0, and MSB of the
+		 * last byte to 1.
+		 * HPRE hardware module uses big-endian mode, so the bytes to be
+		 * set are reversed compared to RFC 7748
+		 * It is a constraint of HPRE hardware that key_bytes will be set
+		 * to 576 when curve_bytes is between 384 and 576, and the high-order
+		 * bits will be set to 0.
+		 */
+		dst[55 + bsize - dsize] &= 0xFC;
+		dst[0 + bsize - dsize] |= 0x80;
+	}
+
+	shift = bsize - psize;
```

Comment:
How would you describe `shift`, what does it represent?

Reply:
Shift means the offset I need to take into accout when I deal with the data related to protocol

Reply:
due to hardware limit mentioned above

Reply:
It sounds like `offset` would be a better name. Please add a comment describing how this variable is used.

Reply:
solved